ide/git/src/org/netbeans/modules/git/ui/history/DefaultGitHyperlinkProvider.java

```diff
@@ -0,0 +1,130 @@
+import org.netbeans.libs.git.GitRemoteConfig;
+import org.netbeans.libs.git.GitRepository;
+import org.netbeans.libs.git.GitURI;
+import org.netbeans.libs.git.progress.ProgressMonitor.DefaultProgressMonitor;
+import org.netbeans.modules.versioning.util.VCSHyperlinkProvider;
+import org.openide.awt.StatusDisplayer;
+import org.openide.util.Exceptions;
+import org.openide.util.NbBundle.Messages;
+import org.openide.util.RequestProcessor;
+import org.openide.util.lookup.ServiceProvider;
+
+/**
+ * Looks for '#123' patterns in text (e.g commit messages) and tries to link
+ * them to a fitting web page.
+ * 
+ * <p>The repository host is identified by inspecting the registered remotes
+ * of the local git repository.
+ * 
+ * @author mbien
+ */
+@Messages({
+    "# {0} - issue ID",
+    "TT_DefaultGitHyperlinkProvider.link=Click to open {0} in Browser",
+    "MSG_DefaultGitHyperlinkProvider.link.failed=Could not resolve issue link."
+})
+@ServiceProvider(service=VCSHyperlinkProvider.class, position = Integer.MAX_VALUE)
+public class DefaultGitHyperlinkProvider extends VCSHyperlinkProvider {
+    
+    private static final Pattern ids = Pattern.compile("#[0-9]+");
+
+    @Override
+    public int[] getSpans(String text) {
+        return ids.matcher(text).results()
+                                .flatMapToInt(r -> IntStream.of(r.start(), r.end()))
+                                .toArray();
+    }
+
+    @Override
+    public String getTooltip(String text, int offsetStart, int offsetEnd) {
+        return Bundle.TT_DefaultGitHyperlinkProvider_link(text.substring(offsetStart, offsetEnd));
+    }
+
+    @Override
+    public void onClick(File file, String text, int offsetStart, int offsetEnd) {
+        Path path = file.toPath();
+        while (!Files.isDirectory(path) || Files.notExists(path.resolve(".git"))) {
+            path = path.getParent();
+            if (path == null) {
+                return;
+            }
+        }
+        Path repo = path;
+        RequestProcessor.getDefault().post(() -> {
+            openIssue(repo, text.substring(offsetStart + 1, offsetEnd));
+        });
+    }
+
+    private static void openIssue(Path repo, String id) {
+        try (GitClient client = GitRepository.getInstance(repo.toFile()).createClient()) {
+            Map<String, GitRemoteConfig> remotes = client.getRemotes(new DefaultProgressMonitor());
+            GitRemoteConfig origin =
+                    remotes.getOrDefault("origin4nb",
+                        remotes.getOrDefault("upstream",
+                            remotes.getOrDefault("origin",
+                                remotes.getOrDefault("origin/HEAD",
+                                    remotes.values().stream()
+                                                    .filter(r -> r.getRemoteName().startsWith("origin/"))
+                                                    .findFirst()
+                                                    .orElse(null)
+            ))));
```

Comment:
I just looked through the code again this morning with a fresh mind and even though this is fairly compact, it isn't ideal.

the `default` part is always evaluated since its not a lambda. I am going to move the stream out of the tree into a separate `if origin==null` check.

only a minor change, going to merge after that.